DTB only returns top-level package references

drewnoakes opened the pull request:
Historically the full graph of package references has been returned in design-time builds and used to populate the Dependencies node in Solution Explorer.

This graph can be very large. Many sessions won't expand the Dependencies node, so we would like to avoid this up-front cost during solution load. This will be achieved by lazily retrieving transitive dependency information from the assets file directly when needed.

This PR modifies `PreprocessPackageDependenciesDesignTime` in the following ways:

- Only return items of type `Package` (exclude others like diagnostics, targets, assemblies)
- Don't return items for all target (only those in the current target framework are required, as we run a DTB per target)
- Simplify the IDs used for returned items to reduce allocations and simplify processing on the Project System side
- Remove intermediate objects (via the private `ItemMetadata` type hierarchy) to reduce allocations

The corresponding Project System changes are able to identify when the SDK is operating with this new behaviour or the old (legacy) behaviour and adapt accordingly.

---

Questions:

In `Microsoft.PackageDependencyResolution.targets`:

1. Can we remove the `ReportAssetsLogMessages` task? It no longer seems necessary for our purposes, but someone theoretically may depend upon it.

2. In this task, outputs `TargetDefinitions`, `FileDefinitions` and `FileDependencies` appear to be no longer used. Again, is it safe to remove them and the code that allocates them?

```xml
    <ResolvePackageDependencies
      ProjectPath="$(MSBuildProjectFullPath)"
      ProjectAssetsFile="$(ProjectAssetsFile)"
      ProjectLanguage="$(Language)"
      ContinueOnError="ErrorAndContinue">
      <Output TaskParameter="TargetDefinitions" ItemName="TargetDefinitions" />
      <Output TaskParameter="PackageDefinitions" ItemName="PackageDefinitions" />
      <Output TaskParameter="FileDefinitions" ItemName="FileDefinitions" />
      <Output TaskParameter="PackageDependencies" ItemName="PackageDependencies" />
      <Output TaskParameter="FileDependencies" ItemName="FileDependencies" />
    </ResolvePackageDependencies>
```

drewnoakes committed:
Remove unused property setters

drewnoakes committed:
Update documentation

drewnoakes committed:
Simplify lazy initialization

drewnoakes committed:
Make methods static

drewnoakes committed:
Reduce string allocations

String.Replace(char, char) allocates a new string even if no replacement is made.

drewnoakes committed:
Avoid array covariance

drewnoakes committed:
Remove redundant assignments

drewnoakes committed:
Array cannot be null or have negative item length

drewnoakes committed:
Inline out var declarations

drewnoakes committed:
Reduce constructor visibilities

drewnoakes committed:
Use local function

drewnoakes committed:
Formatting

drewnoakes committed:
Remove diagnostics from PreprocessPackageDependenciesDesignTime

These are no longer being consumed by the Dependencies node, which now
sources them directly from the assets file.

drewnoakes committed:
Documentation

drewnoakes committed:
Make local functions static

drewnoakes committed:
Convert property to local variable

drewnoakes committed:
Replace O(N) with O(1) lookup

drewnoakes committed:
Reduce allocation when applying item metadata

drewnoakes committed:
Remove assemblies from PreprocessPackageDependenciesDesignTime

From now on, this task will only return top-level items, so we do not
need to output assemblies, nor do we need to accept FileDefinitions.

drewnoakes committed:
Remove non-top-level packages from PreprocessPackageDependenciesDesignTime

Performance work on the Dependencies node in dotnet/project-system mean
that only top-level package data is needed from the design-time build.

This commit removes items and metadata related to transitive
dependencies and hierarchy.

- Remove metadata Dependencies, which used to list child IDs
- Remove metadata IsTopLevelDependency, which is now implicitly true
- Remove 'target' items
- Remove object cloning, now that instances are immutable

drewnoakes committed:
Filter DTB packages by TargetFrameworkMoniker

Previously PreprocessPackageDependenciesDesignTime returned package
dependencies for all targets, and PackageRuleHandler in the project
system would filter based on TargetFrameworkMoniker.

This change applies the filtering in this task instead.

As part of this change we are now able to remove the target prefix from
the item spec of output items. This reduces string allocation, improves
the chance of pooling and reduces string manipulation by the consumer.

drewnoakes committed:
Don't set IsTopLevelDependency metadata

This metadata always has 'true' as its value. We can infer this in the
project system.

drewnoakes committed:
Don't set Type metadata

This metadata always has 'Package' as its value. We can infer this in
the project system.

drewnoakes committed:
Require Name for DTB resolved package reference items

drewnoakes committed:
Extract var for common subexpression

drewnoakes committed:
Update unit tests for PreprocessPackageDependenciesDesignTime

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -97,11 +97,7 @@ protected override void ExecuteCore()
             DependenciesDesignTime = DependenciesWorld.Select(itemKvp =>
             {
                 var newTaskItem = new TaskItem(itemKvp.Key);
-                foreach (var metadataKvp in itemKvp.Value.ToDictionary())
```
nit: Did you do a profile on this? What is the improvement?

In general I don't like perf change without tests to lock the perf improvement to prevent future regression or at least some profile it to show visible changes. This change is ok considering it is equality readable.

drewnoakes replied:
This was mostly about making the code clearer. The reduction in allocation is a bonus.

I'm interested to know how you lock perf improvements in with tests.

wli3 reviewed (commented):
Some nit in the beginning. And then I am lost at https://github.com/dotnet/sdk/pull/11358/commits/1baaca226428ae2eebe3023c9f9250bbb728b860 . 

Looks like tests are start from fresh. So I don't know what is actually removed. We should fine a time to look into this change together

wli3 reviewed (approved):
Had an about 1 hour meeting with Drew. Now I understand this change. The only consumer of this code is project system. So removing all these information should be safe. The code left is also straightforward. The new tests around the code are also sufficient to cover and to understand what the code is doing.

wli3 commented:
@dsplaisted do you want to have a look at this PR since this it is rather big? I think it is good after talking to Drew. We plan to merge this in 1 day.

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -269,12 +269,7 @@ private void PopulateDiagnosticsMap()
-                Enum.TryParse(dependencyTypeString, /* ignoreCase */ true, out dependencyType);
-            }
-
+            Enum.TryParse(dependencyTypeString, ignoreCase: true, out DependencyType dependencyType);
```
It's the same: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoa0gZmoG9qACNlmPHAWxYFEWjFgEE0LAEIsAvq3ay2pEvMHL2/LtwB0AFSgBPAAoBDKAGcYACgBEVsRBwZ+HMgEoA3KvZ8NO/cbOWXAA2QXYOTjAk7p4KZACcFjCuHlRqakTxiVEpajJUUkA=

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -704,7 +704,7 @@ internal static HashSet<string> GetImplicitPackageReferences(string defaultImpli
 
             var packageNames = defaultImplicitPackages.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
-            if (packageNames == null || packageNames.Length <= 0)
+            if (packageNames.Length == 0)
```
`Split` will not return `null`. It can return an empty array.

wli3 commented:
Just learnt Daniel is OOF today. Could we wait until end of the Friday?

dsplaisted commented:
We should also update the [minimum MSBuild version](https://github.com/dotnet/toolset/blob/release/3.1.4xx/src/redist/minimumMSBuildVersion) to 16.7 for this release, right?

dsplaisted reviewed (approved):
Looks good.

Should we also update the [minimum MSBuild version](https://github.com/dotnet/toolset/blob/release/3.1.4xx/src/redist/minimumMSBuildVersion) to 16.7 for this release?  Otherwise using this SDK with an older VS will have a broken dependency tree, right?

To Drew's questions:

Let's remove the `ReportAssetsLogMessages` task.

I think the `TargetDefinitions`, `FileDefinitions`, and `FileDependencies` may be used by some of our build infrastructure (in dotnet/installer, for example).  However, I would propose that by default we don't produce them, with a flag that can be used to turn them back on.

@KathleenDollard, the two previous changes are ones we should track / document as part of our "possibly breaking changes for MSBuild logic" process.

drewnoakes committed:
Remove ReportAssetsLogMessages task

Since VS 16.7, design-time builds no longer consume assets file
diagnostics as items. This commit removes their generation and
supporting code.

drewnoakes committed:
Make methods static

drewnoakes committed:
Discard unused out parameters

drewnoakes committed:
Inline methods

drewnoakes committed:
ResolvePackageDependencies produces fewer outputs

From VS 16.7 onwards, the DTB no longer requires items of type
`TargetDefinitions`, `FileDefinitions`, `FileDependencies`.

This commit stops producing them by default. To restore the previous
behaviour, set the `EmitLegacyAssetsFileItems` property to `true`.

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -2,702 +2,139 @@
-    /// dependency graph.
-    /// If any changes are made here, make sure corresponding changes are made to NuGetDependenciesSubTreeProvider
-    /// in roslyn-project-system repo and corresponding tests.
+    /// Filters and projects items produced by <see cref="ResolvePackageDependencies"/> for consumption by
```
Meant as the verb "projects", as in "performs a projection".

drewnoakes commented:
@dsplaisted thanks for your review.

> Should we also update the minimum MSBuild version to 16.7 for this release? Otherwise using this SDK with an older VS will have a broken dependency tree, right?

Yes, using this SDK with older VS versions will break. Submitted https://github.com/dotnet/toolset/pull/4542 with the update.

> Let's remove the ReportAssetsLogMessages task.

Removed in ab973892824c132a4eeeab01cb371e1541c41a1f. Please review.

> I think the `TargetDefinitions`, `FileDefinitions`, and `FileDependencies` may be used by some of our build infrastructure (in dotnet/installer, for example). However, I would propose that by default we don't produce them, with a flag that can be used to turn them back on.

Added flag `EmitLegacyAssetsFileItems` in the most recent commit, but it's causing test failures in CI and I cannot get tests to run locally. I'll pick this up tomorrow, but any guidance would be appreciated. Or we can split those changes out into a follow up PR.

drewnoakes commented:
Fixed test failures and believe this is good to merge.

Please review f86860682c3a185e51f0697a518bf29841108fa5,  28e790043a93334c857a7daf0748b3b1a2e0d062 and https://github.com/dotnet/toolset/pull/4542.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -48,6 +48,10 @@ Copyright (c) .NET Foundation. All rights reserved.
 
     <EmitAssetsLogMessages Condition="'$(EmitAssetsLogMessages)' == ''">true</EmitAssetsLogMessages>
+
+    <!-- Setting this property restores pre-16.7 behaviour of ResolvePackageDependencies to produce
```
```suggestion
    <!-- Setting this property to true restores pre-16.7 behaviour of ResolvePackageDependencies to produce
```

drewnoakes committed:
Document property value

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>
